File: leetcode/2246-longest-path-with-different-adjacent-characters/solution.py
from typing import List


class Solution:
    def longestPath(self, parent: List[int], s: str) -> int:
        n = len(parent)
        children = [[] for _ in range(n)]
        for v in range(1, n):
            children[parent[v]].append(v)

        longest_down = [1] * n  # longest valid downward chain starting at v
        answer = 1

        # Iterative post-order: children must be fully processed (their
        # longest_down finalized) before their parent combines the results,
        # so recursion depth up to 10^5 is avoided with an explicit stack.
        order = []
        stack = [0]
        while stack:
            u = stack.pop()
            order.append(u)
            for c in children[u]:
                stack.append(c)

        for u in reversed(order):
            best1 = best2 = 0
            for c in children[u]:
                if s[c] != s[u]:
                    chain = longest_down[c]
                    if chain > best1:
                        best1, best2 = chain, best1
                    elif chain > best2:
                        best2 = chain

            longest_down[u] = 1 + best1
            answer = max(answer, 1 + best1 + best2)

        return answer

File: leetcode/2246-longest-path-with-different-adjacent-characters/solution.cpp
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

class Solution {
public:
    int longestPath(vector<int>& parent, string s) {
        int n = parent.size();
        vector<vector<int>> children(n);
        for (int v = 1; v < n; v++) {
            children[parent[v]].push_back(v);
        }

        vector<int> longestDown(n, 1);
        int answer = 1;

        vector<int> order;
        vector<int> stack = {0};
        while (!stack.empty()) {
            int u = stack.back();
            stack.pop_back();
            order.push_back(u);
            for (int c : children[u]) {
                stack.push_back(c);
            }
        }

        for (int i = (int)order.size() - 1; i >= 0; i--) {
            int u = order[i];
            int best1 = 0, best2 = 0;
            for (int c : children[u]) {
                if (s[c] != s[u]) {
                    int chain = longestDown[c];
                    if (chain > best1) {
                        best2 = best1;
                        best1 = chain;
                    } else if (chain > best2) {
                        best2 = chain;
                    }
                }
            }
            longestDown[u] = 1 + best1;
            answer = max(answer, 1 + best1 + best2);
        }

        return answer;
    }
};
